src/access/Subjects.h

```diff
@@ -0,0 +1,208 @@
+        }
+        else if (mSubject.Is<NodeSubject>())
+        {
+            return mSubject.Get<NodeSubject>().nodeId;
+        }
+        else if (mSubject.Is<GroupSubject>())
+        {
+            return static_cast<uint64_t>(mSubject.Get<GroupSubject>().groupId) << 48;
+        }
+        else
+        {
+            return std::numeric_limits<uint64_t>::max();
+        }
+    }
+
+    bool operator==(const ScopedSubject & that) const { return this->mSubject == that.mSubject; }
+
+private:
+    Variant<PaseSubject, NodeSubject, GroupSubject> mSubject;
+};
+
+/**
+ * A subject is a global unique identifier. It suite 2 purpose:
+ *
+ *  1. Identify an entity. operator== can be used to check if 2 subjects are identical.
+ *  2. Associate to ACL entries to grant privileges
+ */
+class Subject
+{
+public:
+    Subject() : mFabricIndex(kUndefinedFabricIndex) {}
+
+    template <typename T, class... Args>
+    constexpr explicit Subject(FabricIndex fabricIndex, InPlaceTemplateType<T>, Args &&... args) :
+        mFabricIndex(fabricIndex), mScopedSubject(InPlaceTemplate<T>, std::forward<Args>(args)...)
+    {}
+
+    template <typename T, typename... Args>
+    static Subject Create(FabricIndex fabricIndex, Args &&... args)
+    {
+        return Subject(fabricIndex, InPlaceTemplate<T>, std::forward<Args>(args)...);
+    }
+
+    static Subject CreatePaseSubject()
+    {
+        // Return a unique subject for all PASE sessions.
+        return Subject(kUndefinedFabricIndex, InPlaceTemplate<PaseSubject>);
+    }
+
+    FabricIndex GetFabricIndex() const { return mFabricIndex; }
+    const ScopedSubject & GetScopedSubject() const { return mScopedSubject; }
+
+    bool operator==(const Subject & that) const
+    {
+        return this->mFabricIndex == that.mFabricIndex && this->mScopedSubject == that.mScopedSubject;
+    }
+
+private:
+    FabricIndex mFabricIndex;
+    ScopedSubject mScopedSubject;
+
+    friend bool operator==(const Subject &, const OperationalNodeId &);
+};
+
+/**
+ * OperationalNodeId identifies an individual Node on a Fabric. It is a special type of Subject targeting to NodeSubject. It is
+ * interchangeable with the generic Subject type but uses less memory.
+ */
+class OperationalNodeId
+{
+public:
+    OperationalNodeId(FabricIndex fabricIndex, NodeId nodeId) : mFabricIndex(fabricIndex), mNodeId(nodeId) {}
+    Subject ToSubject() { return Subject::Create<NodeSubject>(mFabricIndex, mNodeId); }
+
+    FabricIndex GetFabricIndex() const { return mFabricIndex; }
+    NodeId GetNodeId() const { return mNodeId; }
+
+private:
+    FabricIndex mFabricIndex;
+    NodeId mNodeId;
+
+    bool operator==(const OperationalNodeId & that) const
+    {
+        return this->mFabricIndex == that.mFabricIndex && this->mNodeId == that.mNodeId;
+    }
+
+    friend bool operator==(const Subject &, const OperationalNodeId &);
+};
+
+inline bool operator==(const Subject & subject, const OperationalNodeId & node)
```

Comment:
Do we need one like this but with the order reversed too?

But also, defining global operators in namespaces is usually a path to pain. What we should probably do is just define actual class-method `operator==` on both classes that take the right types.

Reply:
Define member function `operator==` for 2 different types are more pain than non-member. If a binary operator is defined as a class member, implicit conversions will apply to the right-hand argument. Which is not expect here.

I'll add the reversed one too.

Reply:
> Define member function operator== for 2 different types are more pain than non-member.

No, it's really not.  The problem with namespaced (i.e. not in the global namespace) non-member operator== is that you get various ambiguity errors if you aren ot in the exact namespace where it's defined.

As far as implicit conversions go... why are there any implicit conversions involved here?

Reply:
Anyway, `operator==` is not being used anymore. Because for CATs matching, the version compare is not symmetric.